content/blog/2022-05-19-fast-brili/index.md

```diff
@@ -39,6 +39,51 @@ very compressed representation of Bril instructions: (almost) all of them are
 represented in 64 bits, with some taking multiple 64 bit words as needed. See
 the [documentation](https://github.coecis.cornell.edu/cs897/fast-bril/blob/master/doc/brb.pdf).
 
+The first component of the bytecode translation is variable names. We translate
+variables to 16 bit numbers in the order that they appear in the function. For
+functions with arguments, the first argument is 0, then 1 et cetera, and the
+variables used in the body follow. This allows the following encoding of "most"
+instructions:
+
+<p align="center">
+	<img src="insn-layout.png"/>
+</p>
+
+By "most" instructions we mean the instructions of the form `t1 = t2 op
+t3`. `arg1`, `arg2`, and `arg3` are the aforementioned numbers of the
+variables.
+
+We also have to translate labels, as strings don't fit into our representation
+very well. We will have the instructions of a function in an array of 64 bit
+bytecode instructions, so we represent labels as indices in this array. We also
+have to keep track of which instructions had a label in the original program so
+that we can interpret the φ instruction efficiently.
+
+We notice that nearly all operations are coincidentally encodings of type
+information. For example, `add` only operates on `int`
+operands/results. However, the `id` instruction is an exception to this, but it
+only has one argument so we encode the type in the space for the second
+argument. See the more detailed docs for the type encoding.
```

Comment:
Interesting! It also occurs to me that it might have been possible to just not record the type in the `id` at all because it doesn't affect how the program executes… except that then you couldn't recover the original Bril JSON, at least not without doing some global type inference.

Reply:
this is true for the `id` instruction, so perhaps it was a bad choice. I originally didn't have this, but then thought i would want it for asm generation (floats need different registers), but then realized that if i was just moving a float from memory to memory, this can go through a general register. This might be useful info to have around for register allocation, but im not sure.

The `call` instruction  on the other hand DOES need this information to be conformant to most calling conventions, as float args are not passed through general purpose registers, so maybe that would have been a better example

If we don't record the type in `id`, then we have no use for those 16 bits, so other than making the translation a teensy bit more expensive, there is nothing lost by recording it

Reply:
Huh, interesting. It would definitely be a bigger change, but one option would be to associate a variable->type table with every function (because Bril [disallows changing the type of a variable within a given function](https://capra.cs.cornell.edu/bril/lang/wellformed.html)) and then just let instructions look it up there when they need it.